### PR TITLE
Always clone the nodes on the path to the write ref point

### PR DESCRIPTION
**Pull request by platfowner**
Summary:
- Always clone the nodes on the path to the write ref point
- Add tests

Testing:
- Passed unit tests `yarn run test_unit`
- Passed integration tests `yarn run test_integration`

#297

### CODE REVIEW

**Comment by minsulee2** [comment]
LGTM!

**Review by minsulee2 — APPROVED** [review]

File: unittest/db.test.js [hunk 1]
```diff
@@ -2847,6 +2847,140 @@ describe("Test proof with database", () => {
...
+      const child2 = node.db.stateRoot.getChild('values').getChild('test').getChild('child_2');
+      const child21 = child2.getChild('child_21');
+
+      expect(node.db.getRefForWriting(['values', 'test', 'child_2', 'child_21']))
+          .to.not.equal(null);
+
+      // The nodes on the path are cloned.
+      const newChild2 = node.db.stateRoot.getChild('values').getChild('test').getChild('child_2');
+      const newChild21 = newChild2.getChild('child_21');
+      expect(newChild2 === child2).to.equal(false);
```
Comment (liayoo) [inline unittest/db.test.js:2939]: question) could you help me understand why these are false in this case?

Comment (platfowner) [reply unittest/db.test.js]: Good question!
Let me share a typical case:
Typical case:
- root_a has subtree child_1a -> child_2 -> child_3 
- root_b has subtree child_1b -> child_2 -> child_3 (child_2 and child_3 are shared)
- Want to change child_3 -> child_3' from root_a

Current behavior:
- Only child_3 is replaced by child_3'
- Reference to child_2 from root_b is also affected

New behavior:
- Shared nodes child_2 is cloned to child_2'
- Reference from root_a: child_1a -> child_2' -> child_3'
- Reference from root_b: child_1b -> child_2 -> child_3 (not affected)

I'll leave some notes.
Please let me know if you have any other ideas.

Comment (liayoo) [reply unittest/db.test.js]: Thank you for such a detailed explanation! I get it now 👍

**Review by liayoo — APPROVED** [review]
LGTM, thanks!!

**Comment by platfowner** [comment]
Thanks for the reviews. Now merging..

**Pull request merged**